File: apps/posting/posting_core/clients/devto.py
from __future__ import annotations

import json

from posting_core.http_client import HttpRequestError, request, request_json
from posting_core.publish_config import DEVTO_API_KEY, log

DEVTO_BASE_HEADERS = {
    "Content-Type": "application/json",
    "User-Agent": "Mozilla/5.0 (compatible; alexgetman-posting/1.0; +https://alexgetman.com)",
}


def _devto_headers() -> dict[str, str]:
    return {**DEVTO_BASE_HEADERS, "api-key": DEVTO_API_KEY}


def publish_to_devto(
    title: str,
    body_markdown: str,
    canonical_url: str | None = None,
    tags: list[str] | None = None,
    main_image: str | None = None,
    published: bool = True,
) -> str | None:
    """
    Publish or create a draft article on dev.to.
    Returns the article URL on success, None on failure.

    canonical_url should point to the original post on alexgetman.com so
    search engines treat that as the authoritative source.
    """
    if not DEVTO_API_KEY:
        log("dev.to API key missing, skipping")
        return None

    # dev.to accepts max 4 tags, lowercase, no spaces
    clean_tags = []
    for t in (tags or []):
        clean = t.lower().replace(" ", "").replace("-", "")[:20]
        if clean:
            clean_tags.append(clean)
    clean_tags = clean_tags[:4]

    article: dict = {
        "title": title,
        "body_markdown": body_markdown,
        "published": published,
    }
    if canonical_url:
        article["canonical_url"] = canonical_url
    if main_image:
        article["main_image"] = main_image
    if clean_tags:
        article["tags"] = clean_tags

    try:
        data = request_json(
            "https://dev.to/api/articles",
            payload={"article": article},
            headers=_devto_headers(),
            method="POST",
            timeout=30,
        )
        url = data.get("url")
        log(f"dev.to article published: {url}")
        return url
    except HttpRequestError as exc:
        log(f"dev.to publish failed: {exc.status} {exc.body}")
        return None
    except Exception as exc:
        log(f"dev.to publish error: {exc}")
        return None


def update_devto_article(article_id: int, **kwargs) -> bool:
    """Update an existing dev.to article by ID."""
    if not DEVTO_API_KEY:
        return False
    try:
        request(
            f"https://dev.to/api/articles/{article_id}",
            data=json.dumps({"article": kwargs}).encode("utf-8"),
            headers=_devto_headers(),
            method="PUT",
            timeout=30,
        )
        return True
    except Exception as exc:
        log(f"dev.to update failed: {exc}")
        return False

File: apps/backend/src/social/devto.ts
import type { BackendConfig } from "../config.js";
import { HttpPublishError, type PublishResult } from "../publishing/errors.js";
import { payloadCanonicalUrl, payloadMedia } from "./payload.js";

type DevtoArticleInput = {
  title: string;
  bodyMarkdown: string;
  canonicalUrl?: string | null;
  tags?: string[] | null;
  mainImage?: string | null;
  published?: boolean;
};

export function devtoArticleFromPayload(payload: Record<string, unknown>, config: BackendConfig): DevtoArticleInput {
  const title =
    stringValue(payload.title) ||
    firstLine(stringValue(payload.text_en) || stringValue(payload.text) || stringValue(payload.bodyMarkdown)) ||
    "Alex Getman update";
  const bodyMarkdown =
    stringValue(payload.bodyMarkdown) ||
    stringValue(payload.body_markdown) ||
    stringValue(payload.text_en) ||
    stringValue(payload.text) ||
    "";
  const canonicalUrl = payloadCanonicalUrl(payload, config);
  const tags = Array.isArray(payload.tags) ? payload.tags.map((tag) => String(tag)) : [];
  const mainImage = stringValue(payload.mainImage) || stringValue(payload.main_image) || null;
  const inlineImage = mainImage ?? payloadMedia(payload).find((item) => item.type === "IMAGE" && item.vpsUrl)?.vpsUrl ?? null;
  return {
    title,
    bodyMarkdown:
      inlineImage && !bodyMarkdown.includes(`](${inlineImage})`) ? `![${title}](${inlineImage})\n\n${bodyMarkdown}` : bodyMarkdown,
    canonicalUrl,
    tags,
    // Dev.to needs this field for the card cover; the same image is also put
    // into Markdown so the article body remains complete outside the card view.
    mainImage: inlineImage,
    published: payload.published == null ? true : Boolean(payload.published),
  };
}

export async function publishToDevto(
  input: DevtoArticleInput,
  config: BackendConfig,
  fetchImpl: typeof fetch = fetch,
): Promise<PublishResult> {
  if (!config.DEVTO_API_KEY) {
    return { skipped: true, reason: "missing DEVTO_API_KEY" };
  }
  const article: Record<string, unknown> = {
    title: input.title,
    body_markdown: input.bodyMarkdown,
    published: input.published ?? true,
  };
  if (input.canonicalUrl) article.canonical_url = input.canonicalUrl;
  if (input.mainImage) article.main_image = input.mainImage;
  const tags = cleanTags(input.tags ?? []);
  if (tags.length > 0) article.tags = tags;

  const response = await fetchImpl("https://dev.to/api/articles", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "User-Agent": "Mozilla/5.0 (compatible; alexgetman-backend/1.0; +https://alexgetman.com)",
      "api-key": config.DEVTO_API_KEY,
    },
    body: JSON.stringify({ article }),
  });
  const body = await response.text();
  if (!response.ok) {
    throw new HttpPublishError(`dev.to publish failed: ${response.status} ${body}`, response.status, body);
  }
  const data = body ? (JSON.parse(body) as Record<string, unknown>) : {};
  const url = typeof data.url === "string" ? data.url : null;
  return { ok: true, id: typeof data.id === "number" || typeof data.id === "string" ? data.id : url, url, raw: data };
}

export async function updateDevtoArticle(
  articleId: number,
  patch: Partial<DevtoArticleInput>,
  config: BackendConfig,
  fetchImpl: typeof fetch = fetch,
): Promise<boolean> {
  if (!config.DEVTO_API_KEY) return false;
  const article: Record<string, unknown> = {};
  if (patch.title) article.title = patch.title;
  if (patch.bodyMarkdown) article.body_markdown = patch.bodyMarkdown;
  if (patch.canonicalUrl) article.canonical_url = patch.canonicalUrl;
  if (patch.mainImage) article.main_image = patch.mainImage;
  if (patch.tags) article.tags = cleanTags(patch.tags);
  if (patch.published != null) article.published = patch.published;
  const response = await fetchImpl(`https://dev.to/api/articles/${articleId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "User-Agent": "Mozilla/5.0 (compatible; alexgetman-backend/1.0; +https://alexgetman.com)",
      "api-key": config.DEVTO_API_KEY,
    },
    body: JSON.stringify({ article }),
  });
  if (!response.ok) {
    const body = await response.text();
    throw new HttpPublishError(`dev.to update failed: ${response.status} ${body}`, response.status, body);
  }
  return true;
}

function cleanTags(tags: string[]): string[] {
  return tags
    .map((tag) => tag.toLowerCase().replaceAll(" ", "").replaceAll("-", "").slice(0, 20))
    .filter(Boolean)
    .slice(0, 4);
}

function stringValue(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

function firstLine(value: string): string {
  return value.split(/\r?\n/, 1)[0]?.trim() ?? "";
}
